Loading required libraries and dataset. Since the dataset is huge, we sample only 2% of it in our study. 
  

In [1]:
import pandas as pd
import numpy as np
pd.set_option("display.max_rows",None,"display.max_columns", None)
import warnings
warnings.filterwarnings('always') 

Column graphical_appearance_name correspond to product_pattern and perceived_colour_value_name corresponds to product_color to make the interpretation easier for the reader.


In [2]:
hmdata = pd.read_csv('C:/Users/Study/OneDrive/Desktop/DU/Business_Intelligence/Home_assignment_BI/dataset_hm.csv')
hmdata = hmdata.sample(frac=0.2, random_state=100)
hmdata.describe(include=object)
hmdata.shape
hmdata.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 15971 entries, 65748 to 24132
Data columns (total 14 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   t_date                       15971 non-null  object 
 1   product_type_no              15971 non-null  int64  
 2   product_type_name            15971 non-null  object 
 3   product_group_no             15971 non-null  int64  
 4   product_group_name           15971 non-null  object 
 5   graphical_appearance_no      15971 non-null  int64  
 6   graphical_appearance_name    15971 non-null  object 
 7   colour_group_code            15971 non-null  int64  
 8   colour_group_name            15971 non-null  object 
 9   perceived_colour_value_id    15971 non-null  int64  
 10  perceived_colour_value_name  15971 non-null  object 
 11  index_group_no               15971 non-null  int64  
 12  index_group_name             15971 non-null  object 
 13  Price       

In [3]:
!pip install sklearn
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report
from sklearn import metrics 
from sklearn.metrics import confusion_matrix
!pip install statsmodels 
import matplotlib.pyplot as plt


You should consider upgrading via the 'c:\users\study\onedrive\desktop\du\business_intelligence\labs\venv\scripts\python.exe -m pip install --upgrade pip' command.
sys:1: ResourceWarning: unclosed socket <zmq.Socket(zmq.PUSH) at 0x11663208820>
sys:1: ResourceWarning: unclosed socket <zmq.Socket(zmq.PUSH) at 0x11663208b20>
c:\users\study\onedrive\desktop\du\business_intelligence\labs\venv\lib\site-packages\IPython\utils\_process_win32.py:124: ResourceWarning: unclosed file <_io.BufferedWriter name=4>
  return process_handler(cmd, _system_body)
c:\users\study\onedrive\desktop\du\business_intelligence\labs\venv\lib\site-packages\IPython\utils\_process_win32.py:124: ResourceWarning: unclosed file <_io.BufferedReader name=5>
  return process_handler(cmd, _system_body)
c:\users\study\onedrive\desktop\du\business_intelligence\labs\venv\lib\site-packages\IPython\utils\_process_win32.py:124: ResourceWarning: unclosed file <_io.BufferedReader name=6>
  return process_handler(cmd, _system_body)


You should consider upgrading via the 'c:\users\study\onedrive\desktop\du\business_intelligence\labs\venv\scripts\python.exe -m pip install --upgrade pip' command.
sys:1: ResourceWarning: unclosed socket <zmq.Socket(zmq.PUSH) at 0x116666c4160>
sys:1: ResourceWarning: unclosed socket <zmq.Socket(zmq.PUSH) at 0x116666c4100>
c:\users\study\onedrive\desktop\du\business_intelligence\labs\venv\lib\site-packages\IPython\utils\_process_win32.py:124: ResourceWarning: unclosed file <_io.BufferedWriter name=4>
  return process_handler(cmd, _system_body)
c:\users\study\onedrive\desktop\du\business_intelligence\labs\venv\lib\site-packages\IPython\utils\_process_win32.py:124: ResourceWarning: unclosed file <_io.BufferedReader name=5>
  return process_handler(cmd, _system_body)
c:\users\study\onedrive\desktop\du\business_intelligence\labs\venv\lib\site-packages\IPython\utils\_process_win32.py:124: ResourceWarning: unclosed file <_io.BufferedReader name=6>
  return process_handler(cmd, _system_body)


We prepare are data to fit the model. y_hmdata is the dependent variable and x_hmdata is the independent variable

In [4]:
y_hmdata = hmdata[["product_group_name"]] 
x_hmdata = hmdata[["graphical_appearance_name","perceived_colour_value_name"]]
x_hmdata = pd.get_dummies(x_hmdata)
x_hmdata.shape


(15971, 37)

In [5]:
x_train,x_test,y_train,y_test=train_test_split(x_hmdata,y_hmdata,test_size=0.25,random_state=100)
print(x_train.shape,y_train.shape,x_test.shape,y_test.shape)

(11978, 37) (11978, 1) (3993, 37) (3993, 1)


In [6]:
#fit model
model1 = LogisticRegression(random_state=0, multi_class='multinomial', penalty='none', solver='saga',max_iter=1000000).fit(x_train, y_train.values.ravel())
preds = model1.predict(x_test)
params = model1.get_params()
print(params)


{'C': 1.0, 'class_weight': None, 'dual': False, 'fit_intercept': True, 'intercept_scaling': 1, 'l1_ratio': None, 'max_iter': 1000000, 'multi_class': 'multinomial', 'n_jobs': None, 'penalty': 'none', 'random_state': 0, 'solver': 'saga', 'tol': 0.0001, 'verbose': 0, 'warm_start': False}


In [7]:
#Calculate odds ratio estimates
np.exp(model1.coef_)


array([[5.37278858e-01, 1.29873949e+00, 1.15835773e-01, 4.67843305e+00,
        4.40323563e+00, 4.67504488e-01, 9.17780427e+00, 9.68380106e+00,
        6.25512941e-01, 2.45892288e+00, 1.07280852e+00, 1.30141406e+00,
        2.50906031e+01, 3.77263600e+00, 7.76303282e-01, 3.84610121e-03,
        1.99787122e+00, 2.54485097e-01, 1.27245479e-01, 8.40943025e-01,
        5.64870109e-01, 1.00967129e+00, 7.12924553e+00, 7.99755747e+00,
        3.03557200e-01, 3.72027206e-01, 1.66135302e-01, 1.25325309e+01,
        6.01530076e-01, 8.30389870e-01, 6.38382384e-01, 8.43731152e-01,
        5.19327233e-01, 9.48394331e-01, 6.62994648e-01, 5.20834419e+00,
        1.38124011e+01],
       [2.22221751e-01, 4.30135916e-01, 2.42475537e-01, 2.84936701e+00,
        6.30961903e+00, 1.26931081e+00, 1.24961353e+01, 5.21227944e+01,
        1.17819655e-01, 7.39932619e-01, 3.47768491e-01, 7.54150503e-01,
        1.04885471e+01, 2.75981813e-01, 1.82133336e+00, 8.59774917e+00,
        8.67514588e-01, 2.86697524e-01,

In [8]:
#Print model parameters
print('Intercept: \n', model1.intercept_)
print('Coefficients: \n', model1.coef_)

Intercept: 
 [ 2.35203529  2.59336839  3.65754015 -5.23089456 -1.00033067 -2.3717186 ]
Coefficients: 
 [[-0.62123803  0.26139417 -2.15558184  1.54296324  1.48233964 -0.76034633
   2.21678799  2.2704545  -0.46918326  0.8997234   0.07028     0.26345141
   3.2224934   1.32777396 -0.25321201 -5.56069532  0.69208222 -1.368513
  -2.06163716 -0.17323137 -0.57115947  0.00962483  1.96420541  2.07913618
  -1.19218522 -0.98878829 -1.79495275  2.52832773 -0.50827874 -0.18585997
  -0.44881783 -0.16992138 -0.65522109 -0.0529849  -0.41098836  1.65026199
   2.62556682]
 [-1.50407952 -0.84365404 -1.41685445  1.04709687  1.8420753   0.23847408
   2.52541942  3.95360237 -2.13860017 -0.30119615 -1.05621828 -0.28216332
   2.35028391 -1.28742031  0.59956885  2.15150044 -0.14212295 -1.24932754
  -2.80353051  1.11313066 -0.13542893  0.13895361  0.85689184  1.90076018
  -1.15531882 -1.53126804 -2.50407449  3.40745049 -1.18058209 -0.45466184
   0.10332673 -0.07408839 -0.54331265 -0.25679656  0.09614673  0.66624

In [9]:
#Create a confusion matrix
confusion_matrix(y_test, preds)


array([[  11,   21,  606,    0,    0,    0],
       [   4,  179,  760,    0,    0,    0],
       [  17,   29, 2046,    0,    2,    0],
       [   1,    0,   13,    0,    0,    0],
       [   1,    0,  221,    0,    0,    0],
       [  10,    0,   72,    0,    0,    0]], dtype=int64)

Predicted values are the column-wise values. Actual values are the row wise values


Total test examples of any class = sum of corresponding row (TP + FN for that class)

TPs are the diagonal values of the matrix 

Total FNs of a class = sum of values of corresponding rows - TPs

Total FPs of a class = sum of corresponding columns - TPs

Total TNs of a class = Sum of all columns and rows - sum of rows and colums of that perticular class


In [10]:
#Accuracy statistics
print('Accuracy Score:', metrics.accuracy_score(y_test, preds))


Accuracy Score: 0.5599799649386427


In [11]:
class_report=classification_report(y_test, preds)
print(class_report)

                    precision    recall  f1-score   support

 Garment Full body       0.25      0.02      0.03       638
Garment Lower body       0.78      0.19      0.31       943
Garment Upper body       0.55      0.98      0.70      2094
         Nightwear       0.00      0.00      0.00        14
             Shoes       0.00      0.00      0.00       222
          Swimwear       0.00      0.00      0.00        82

          accuracy                           0.56      3993
         macro avg       0.26      0.20      0.17      3993
      weighted avg       0.51      0.56      0.45      3993



c:\users\study\onedrive\desktop\du\business_intelligence\labs\venv\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
c:\users\study\onedrive\desktop\du\business_intelligence\labs\venv\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
c:\users\study\onedrive\desktop\du\business_intelligence\labs\venv\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control

From confusion matrix we were able to point out that the product distribution is uneven and that most of the data is present in the first three classes, hence we only evaluate odds ratio for Upper garments (class 1), Lower garment (class 2) and Full body garments (class 3). From table<> we can say that there is indeed similarity in patterns and colours across products categories. The model generates an overall accuracy of 56% with a high f-score of 0.7 for upper garments (Class 1).  \